libparsec/crates/types/src/certif.rs

```diff
@@ -235,8 +251,13 @@ pub struct UserCertificate {
 impl_unsecure_load!(UserCertificate -> CertificateSignerOwned);
 impl_unsecure_dump!(UserCertificate);
 impl_dump_and_sign!(UserCertificate);
+impl_base_load!(UserCertificate);
 
 impl UserCertificate {
+    pub fn check_data_integrity(&self) -> DataResult<()> {
```

Comment:
I think this method should be private: right now it is unclear that this method is used internally by `impl_base_load` (and hence that there is no need to call it manually)

If possible, I'd say we should remove this method entirely and replace it by an inline in the macro call (this way it's very obvious when this code is used):
```rust
// Note the `check_data_integrity:` that helps document what this closure is for
impl_base_load!(RevokedUserCertificate, check_data_integrity: |data: RevokedUserCertificate| -> DataResult<()> { Ok(()) });
```

Note `macro_rules` support this by simply passing the closure as an `$expr`:
```rust
macro_rules! foo {
   (function: $function:expr) => {
       $function(42)
   };
}

fn main() {
    let ret = foo!(function: |i| i * 2);
    assert_eq!(ret, 84);
}
```

If this is not possible (typically the `check_data_integrity` is used after the deserialization step because the object gets modified during its lifetime), then it's still worth it to have an explicit link between the method and the macro:
```rust
impl RevokedUserCertificate {
    /// This object can be modified during its lifetime, hence we provide this method to run
    /// sanity checks once a modification occurs.
    pub fn check_data_integrity(&self) -> DataResult<()> {
        ...
    }
}
impl_base_load!(RevokedUserCertificate, check_data_integrity: RevokedUserCertificate::check_data_integrity);

Reply:
[6731fa0](https://github.com/Scille/parsec-cloud/pull/7431/commits/6731fa051bb816f0bb76b1ff21d27139208df024)

Reply:
> If possible, I'd say we should remove this method entirely and replace it by an inline in the macro call (this way it's very obvious when this code is used): [...]

Isn't it less readable to have the integrity logic defined in the macro rather than in the struct impl?

> then it's still worth it to have an explicit link between the method and the macro

Is it tho? I can do it sure but I don't see the benefit.

Reply:
Here's how it looks: e2c33cc83c1c1cb57fee88a028697086f86adc25

Reply:
> Isn't it less readable to have the integrity logic defined in the macro rather than in the struct impl?

Yes, it gets less readable pretty fast when the integrity logic gets bigger.
Hence the `impl_base_load!(RevokedUserCertificate, check_data_integrity: RevokedUserCertificate::check_data_integrity);` where there is a dedicated `check_data_integrity` method

> Is it tho? I can do it sure but I don't see the benefit.

If the function is private, the reader understand it is an implementation detail.
However if the function is public, the reader assume it is for external usage. The fact the function is used in a macro further hide the fact it is an internal detail.
On top of that we have to consider the fact a lot of other data structures have a similar `check_data_integrity`, which is often to be used outside of deserialization (e.g. `LocaFilelManifest`).

So in the end I'd say:
- If the function is private and the macro define in the same module,  no need to make the link between the two explict.
- Avoid `pub` as much as possible for this method, use `pub(crate)` or `pub(super)` if it is used from another module in the crate. This way we won't be puzzled about this method when used from `libparsec_client`.
- Add a comment to explain why it is not `pub`.

I would add something like:
```rust
/// This structure represent immutable data (as it is created once, signed, and never updated).
/// Hence this `check_data_integrity` is only used during deserialization (and also as sanity check
/// right before serialization) and not exposed publicly.
pub fn check_data_integrity(&self) -> DataResult<()> {
...
}
```

And for structures that can be modified:
```rust
/// This structure represent mutable data (it gets loaded from disk, updated, then stored back modified)
/// Hence this `check_data_integrity`'s main goal is during deserialization.
/// However it is also useful as sanity check:
/// - Right before serialization
/// - After any modification (hence the need for this method to be public)
pub fn check_data_integrity(&self) -> DataResult<()> {
...
}
```